llama_index/node_parser/file/markdown.py

```diff
@@ -0,0 +1,146 @@
+        callback_manager = callback_manager or CallbackManager([])
+
+        return cls(
+            include_metadata=include_metadata,
+            include_prev_next_rel=include_prev_next_rel,
+            callback_manager=callback_manager,
+        )
+
+    @classmethod
+    def class_name(cls) -> str:
+        """Get class name."""
+        return "MarkdownNodeParser"
+
+    def get_nodes_from_documents(
+        self,
+        documents: Sequence[TextNode],
+        show_progress: bool = False,
+    ) -> List[BaseNode]:
+        """Parse document into nodes.
+
+        Args:
+            documents (Sequence[TextNode]): TextNodes or Documents to parse
+
+        """
+        with self.callback_manager.event(
+            CBEventType.NODE_PARSING, payload={EventPayload.DOCUMENTS: documents}
+        ) as event:
+            all_nodes: List[BaseNode] = []
+            documents_with_progress = get_tqdm_iterable(
+                documents, show_progress, "Parsing documents into nodes"
+            )
+
+            for document in documents_with_progress:
+                nodes = self.get_nodes_from_node(document)
+                all_nodes.extend(nodes)
+
+            event.on_end(payload={EventPayload.NODES: all_nodes})
+
+        return all_nodes
+
+    def get_nodes_from_node(self, node: BaseNode) -> List[TextNode]:
+        """Get nodes from document"""
+
+        text = node.get_content(metadata_mode=MetadataMode.NONE)
+        markdown_nodes = []
+        lines = text.split("\n")
+        metadata: dict[str, str] = {}
+
+        current_section = ""
+
+        for line in lines:
+            header_match = re.match(r"^(#+)\s(.*)", line)
+            if header_match:
+                if current_section != "":
+                    markdown_nodes.append(
+                        self._build_node_from_split(
+                            current_section.strip(), node, metadata
+                        )
+                    )
+                metadata = self._update_metadata(
+                    metadata, header_match.group(2), len(header_match.group(1).strip())
+                )
+                current_section = f"{header_match.group(2)}\n"
+            else:
+                current_section += line + "\n"
+
+        markdown_nodes.append(
+            self._build_node_from_split(current_section.strip(), node, metadata)
+        )
+
+        return markdown_nodes
+
+    def _update_metadata(
+        self, headers: dict, new_header: str, new_header_level: int
+    ) -> dict:
+        """Update the markdown headers for metadata
+
+        Removes all headers that are equal or less than the level
+        of the newly found header
+        """
+        updated_headers = {}
+
+        for i in range(1, new_header_level):
+            key = f"Header {i}"
+            if key in headers:
+                updated_headers[key] = headers[key]
+
+        updated_headers[f"Header {new_header_level}"] = new_header
+        return updated_headers
```

Comment:
So if I have a node two levels in, does the metadata look like

```
print(node.metadata)
> {'Header 1: 'Title1', 'Header 2: 'Title 2'}
```

Reply:
Yeah that's right, here's the metadata it creates running it on the LlamaIndex Readme:

```
{'filename': 'README.md', 'extension': '.md', 'Header 1': '🗂️ LlamaIndex 🦙'}
{'filename': 'README.md', 'extension': '.md', 'Header 1': '🗂️ LlamaIndex 🦙', 'Header 3': 'Ecosystem'}
{'filename': 'README.md', 'extension': '.md', 'Header 1': '🗂️ LlamaIndex 🦙', 'Header 2': '🚀 Overview'}
{'filename': 'README.md', 'extension': '.md', 'Header 1': '🗂️ LlamaIndex 🦙', 'Header 2': '🚀 Overview', 'Header 3': 'Context'}
{'filename': 'README.md', 'extension': '.md', 'Header 1': '🗂️ LlamaIndex 🦙', 'Header 2': '🚀 Overview', 'Header 3': 'Proposed Solution'}
{'filename': 'README.md', 'extension': '.md', 'Header 1': '🗂️ LlamaIndex 🦙', 'Header 2': '💡 Contributing'}
{'filename': 'README.md', 'extension': '.md', 'Header 1': '🗂️ LlamaIndex 🦙', 'Header 2': '📄 Documentation'}
{'filename': 'README.md', 'extension': '.md', 'Header 1': '🗂️ LlamaIndex 🦙', 'Header 2': '💻 Example Usage'}
{'filename': 'README.md', 'extension': '.md', 'Header 1': 'rebuild storage context'}
{'filename': 'README.md', 'extension': '.md', 'Header 1': 'load index'}
{'filename': 'README.md', 'extension': '.md', 'Header 1': 'load index', 'Header 2': '🔧 Dependencies'}
{'filename': 'README.md', 'extension': '.md', 'Header 1': 'load index', 'Header 2': '📖 Citation'}
```

Reply:
Fixed the code block headers parsing haha:

```
{'filename': 'README.md', 'extension': '.md', 'Header 1': '🗂️ LlamaIndex 🦙'}
{'filename': 'README.md', 'extension': '.md', 'Header 1': '🗂️ LlamaIndex 🦙', 'Header 3': 'Ecosystem'}
{'filename': 'README.md', 'extension': '.md', 'Header 1': '🗂️ LlamaIndex 🦙', 'Header 2': '🚀 Overview'}
{'filename': 'README.md', 'extension': '.md', 'Header 1': '🗂️ LlamaIndex 🦙', 'Header 2': '🚀 Overview', 'Header 3': 'Context'}
{'filename': 'README.md', 'extension': '.md', 'Header 1': '🗂️ LlamaIndex 🦙', 'Header 2': '🚀 Overview', 'Header 3': 'Proposed Solution'}
{'filename': 'README.md', 'extension': '.md', 'Header 1': '🗂️ LlamaIndex 🦙', 'Header 2': '💡 Contributing'}
{'filename': 'README.md', 'extension': '.md', 'Header 1': '🗂️ LlamaIndex 🦙', 'Header 2': '📄 Documentation'}
{'filename': 'README.md', 'extension': '.md', 'Header 1': '🗂️ LlamaIndex 🦙', 'Header 2': '💻 Example Usage'}
{'filename': 'README.md', 'extension': '.md', 'Header 1': '🗂️ LlamaIndex 🦙', 'Header 2': '🔧 Dependencies'}
{'filename': 'README.md', 'extension': '.md', 'Header 1': '🗂️ LlamaIndex 🦙', 'Header 2': '📖 Citation'}
```